sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubMessageToRow.java

```diff
@@ -90,44 +102,94 @@ public static Builder builder() {
     return new AutoValue_PubsubMessageToRow.Builder();
   }
 
-  @DoFn.ProcessElement
-  public void processElement(ProcessContext context) {
-    try {
-      List<Object> values = getFieldValues(context);
-      context.output(Row.withSchema(messageSchema()).addValues(values).build());
-    } catch (UnsupportedRowJsonException jsonException) {
-      if (useDlq()) {
-        context.output(DLQ_TAG, context.element());
-      } else {
-        throw new RuntimeException("Error parsing message", jsonException);
-      }
+  private PubsubIO.Write<PubsubMessage> writeMessagesToDlq() {
+    PubsubIO.Write<PubsubMessage> write = PubsubIO.writeMessages().to(deadLetterQueue());
+
+    return timestampAttribute() != null
+        ? write.withTimestampAttribute(timestampAttribute())
+        : write;
+  }
+
+  @Override
+  public PCollection<Row> expand(PCollection<PubsubMessage> input) {
+    Schema schema = payloadSchema();
+    PCollectionTuple rows =
+        input.apply(
+            ParDo.of(
+                    useFlatSchema()
+                        ? new FlatSchemaPubsubMessageToRoW(schema)
+                        : new NestedSchemaPubsubMessageToRow(schema))
+                .withOutputTags(
+                    MAIN_TAG,
+                    deadLetterQueue() != null ? TupleTagList.of(DLQ_TAG) : TupleTagList.empty()));
+
+    if (deadLetterQueue() != null) {
+      rows.get(DLQ_TAG).apply(writeMessagesToDlq());
     }
+
+    return rows.get(MAIN_TAG).setRowSchema(messageSchema());
   }
 
   /**
-   * Get values for fields in the same order they're specified in schema, including timestamp,
-   * payload, and attributes.
+   * A {@link DoFn} to convert a flat schema{@link PubsubMessage} with JSON payload to {@link Row}.
    */
-  private List<Object> getFieldValues(ProcessContext context) {
-    Row payload = parsePayloadJsonRow(context.element());
-    return messageSchema().getFields().stream()
-        .map(
-            field ->
-                getValueForField(
-                    field, context.timestamp(), context.element().getAttributeMap(), payload))
-        .collect(toList());
-  }
+  @Internal
+  class FlatSchemaPubsubMessageToRoW extends DoFn<PubsubMessage, Row> {
```

Comment:
As noted in the test, I think these classes should be private

Reply:
Done